modules/raft-client/src/main/java/org/apache/ignite/raft/client/service/impl/RaftGroupServiceImpl.java

```diff
@@ -0,0 +1,231 @@
+
+            return null;
+        });
+    }
+
+    @Override public CompletableFuture<Void> refreshMembers(boolean onlyAlive) {
+        GetPeersRequest req = factory.createGetPeersRequest().setOnlyAlive(onlyAlive).setGroupId(groupId).build();
+
+        CompletableFuture<GetPeersResponse> fut = sendWithRetry(req);
+
+        return fut.thenApply(resp -> {
+            peers = resp.getPeersList();
+            learners = resp.getLearnersList();
+
+            return null;
+        });
+    }
+
+    @Override public CompletableFuture<Void> addPeers(Collection<PeerId> peerIds) {
+        AddPeersRequest.Builder builder = factory.createAddPeersRequest().setGroupId(groupId);
+
+        for (PeerId peerId : peerIds)
+            builder.addPeer(peerId);
+
+        AddPeersRequest req = builder.build();
+
+        CompletableFuture<ChangePeersResponse> future = sendWithRetry(req);
+
+        return future.thenApply(resp -> {
+            peers = resp.getNewPeersList();
+
+            return null;
+        });
+    }
+
+    @Override public CompletableFuture<Void> removePeers(Collection<PeerId> peerIds) {
+        return null;
+    }
+
+    @Override public CompletableFuture<Void> addLearners(Collection<PeerId> learners) {
+        return null;
+    }
+
+    @Override public CompletableFuture<Void> removeLearners(Collection<PeerId> learners) {
+        return null;
+    }
+
+    @Override public CompletableFuture<Void> snapshot(PeerId peerId) {
+        return null;
+    }
+
+    @Override public CompletableFuture<Void> transferLeadership(PeerId newLeader) {
+        return null;
+    }
+
+    @Override public <R> CompletableFuture<R> run(Command cmd) {
+        return sendWithRetry(factory.createUserRequest().setRequest(cmd).setGroupId(groupId).build());
+    }
+
+    @Override public <R> CompletableFuture<R> run(PeerId peerId, ReadCommand cmd) {
+        UserRequest req = factory.createUserRequest().setRequest(cmd).setGroupId(groupId).build();
+
+        return cluster.sendWithResponse(peerId.getNode(), req, timeout);
+    }
+
+    private <R> CompletableFuture<R> sendWithRetry(Object req) {
+        PeerId leader = this.leader;
+
+        CompletableFuture<Void> refreshLeaderFut = leader == null ? refreshLeader() : FINISHED_FUT;
+
+        // TODO retries.
+        return refreshLeaderFut.<UserResponse<R>>
+            thenCompose(peerId -> cluster.sendWithResponse(leader.getNode(), req, timeout)).
+            thenApply(resp -> resp.response());
+    }
+
+    private void info(String msg, Object... params) {
+        LOG.log(System.Logger.Level.INFO, msg, params);
+    }
+
+    private void warn(String msg, Object... params) {
+        LOG.log(System.Logger.Level.WARNING, msg, params);
+    }
+
+    private void error(String msg, Object... params) {
+        LOG.log(System.Logger.Level.ERROR, msg, params);
+    }
+
+    private void error(String msg, Exception e) {
+        LOG.log(System.Logger.Level.ERROR, msg, e);
+    }
```

Comment:
Let's wrap System.Logger and move this to a separate class to keep code clean.
Later, the logger wrapper can be easily refactored.

Reply:
I've introduced LogWrapper and moved it to the core. It delegates to system logger.